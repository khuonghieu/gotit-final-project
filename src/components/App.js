import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { Route, Redirect, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import {
  Header, Button, ToastContainer, Logo,
} from '@gotitinc/design-system';
import ModalContainer from './Modals';
import { fetchUserInfo } from '../actions/users';
import Catalog from './Catalog';
import UserProfile from './User/UserProfile';
import UserSignOut from './User/UserSignOut';
import ItemDetails from './Items/ItemDetails';
import { chooseModal } from '../actions/modals';
import * as constants from '../constants/actions';

export function App({ user, fetchUserInfo, chooseModal }) {
  // Fetch user info on every render
  useEffect(() => {
    if (user.loggedIn) {
      fetchUserInfo();
    }
  }, [fetchUserInfo, user.loggedIn]);

  return (
    <div>
      <div><ModalContainer /></div>
      <ToastContainer />
      <Header fullWidth>
        <Header.Main>
          <Header.Left>
            <Link to="/">
              <Logo name="gotit" variant="original" />
            </Link>
            <UserProfile user={user} />
          </Header.Left>
          {user.loggedIn
            ? (
              <div>
                <Header.Right>
                  <div><UserSignOut /></div>
                </Header.Right>
              </div>
            ) : (
              <div>
                <Header.Right>
                  <Button onClick={() => chooseModal(constants.SIGN_UP_MODAL)}>Sign up</Button>
                  <Button onClick={() => chooseModal(constants.SIGN_IN_MODAL)}>Sign in</Button>
                </Header.Right>
              </div>
            )}
        </Header.Main>
      </Header>
      <Route
        exact
        path="/catalog/:categoryId?"
        component={Catalog}
      />
      <Route exact path="/catalog/:categoryId/:itemId" component={ItemDetails} />
      <Route exact path="/">
        <Redirect to="/catalog" />
      </Route>
    </div>

  );
}

App.propTypes = {
  user: PropTypes.object.isRequired,
  fetchUserInfo: PropTypes.func.isRequired,
  chooseModal: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
  return {
    user: state.user,
  };
}

const mapDispatchToProps = {
  fetchUserInfo,
  chooseModal,
};

export default connect(mapStateToProps, mapDispatchToProps)(App);
