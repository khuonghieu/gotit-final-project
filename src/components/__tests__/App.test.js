import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { App } from '../App';

configure({ adapter: new Adapter() });

describe('components/App.js', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      user: {
        loggedIn: false,
        currentUser: null,
      },
      modal: null,
      onFetchUserInfo: jest.fn(),
    };
  });
  const setup = () => {
    wrapper = shallow(
      <App {...props} />,
    );
  };

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should not fetch user info when not logged in ', () => {
    setup();
    jest.spyOn(React, 'useEffect').mockImplementation((f) => f());
    expect(props.onFetchUserInfo).not.toHaveBeenCalled();
  });
});
