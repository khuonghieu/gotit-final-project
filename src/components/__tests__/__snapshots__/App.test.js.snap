// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/App.js should render correctly 1`] = `ShallowWrapper {}`;
